timemachine/fe/reweighting.py

```diff
@@ -0,0 +1,192 @@
+    # log(mean(values))
+    # = log(sum(values) / len(values))
+    return logsumexp(log_values) - np.log(len(log_values))
+
+
+def estimate_log_z_ratio(log_importance_weights: Array) -> float:
+    # log(mean(importance_weights))
+    return log_mean(log_importance_weights)
+
+
+def one_sided_exp(delta_us: Array) -> float:
+    # delta_us = -log_importance_weights
+    # delta_f  = -log_z_ratio
+    return -estimate_log_z_ratio(-delta_us)
+
+
+def interpret_as_mixture_potential(u_kn: Array, f_k: Array, N_k: Array) -> Array:
+    """Interpret samples from multiple states as if they originate from a *single* state given by this potential.
+
+    Notes
+    -----
+    * This is nicely illustrated in ref [1] as a way to interpret MBAR [2]
+    * Depends on the accuracy of the input free energy estimates (f_k - f_k[0]) for the source states.
+    * This is not the only way to interpret samples from multiple distributions
+        as if they come from a single mixture distribution.
+        In ref [3], alternatives are systematically enumerated.
+        Assuming f_k are exact, the current approach corresponds to N3 in ref [3].
+        (N3 the most computationally expensive of the options in ref [3],
+        since it requires evaluating all k energy functions on every sample,
+        but it is also the lowest variance.)
+
+    References
+    ----------
+    [1] [Shirts, 2017] Reweighting from the mixture distribution as a better way to describe
+        the Multistate Bennett Acceptance Ratio
+        https://arxiv.org/abs/1704.00891
+    [2] [Shirts, Chodera, 2008] Statistically optimal analysis of samples from multiple equilibrium states.
+        J. Chem. Phys. 129:124105, 2008.
+        http://dx.doi.org/10.1063/1.2978177
+    [3] [Elvira+, 2019] Generalized multiple importance sampling
+        https://arxiv.org/abs/1511.03095
+    """
+    n_states, n_samples = u_kn.shape
+    N_k = np.array(N_k)
+    assert f_k.shape == (n_states,)
+    assert np.sum(N_k) == n_samples
+
+    mixture_u_n = -logsumexp(f_k - u_kn.T, b=N_k, axis=1)
+
+    assert mixture_u_n.shape == (n_samples,)
+
+    return mixture_u_n
+
+
+def construct_endpoint_reweighting_estimator(
+    samples_0: Samples,
+    samples_1: Samples,
+    batched_u_0_fxn: BatchedReducedPotentialFxn,
+    batched_u_1_fxn: BatchedReducedPotentialFxn,
+    ref_params: Params,
+    ref_delta_f: float,
+) -> Callable[[Params], float]:
+    """assuming
+    * endpoint samples (samples_0, samples_1)
+    * precise estimate of free energy difference at initial params
+        ref_delta_f ~= f(ref_params, 1) - f(ref_params, 0)
+
+    construct an estimator of f(params, 1) - f(params, 0)
+    """
+    ref_u_0 = batched_u_0_fxn(samples_0, ref_params)
+    ref_u_1 = batched_u_1_fxn(samples_1, ref_params)
+
+    def endpoint_correction_0(params) -> float:
+        """estimate f(ref, 0) -> f(params, 0) by reweighting"""
+        delta_us = batched_u_0_fxn(samples_0, params) - ref_u_0
```

Comment:
Interface -- Currently the signature is `BatchedReducedPotentialFxn = Callable[[Samples, Params], Energies]`. It may be more general to switch to `BatchedReducedPotentialFxn = Callable[[Params], Energies]`.

Since `samples` will be held constant, a summary of the `samples` can be precomputed once and incorporated into the definition of the batched potential function in some performance optimizations, as in https://github.com/proteneer/timemachine/blob/bb5c7288054ace37804d6c1e3c5cce24ad821d99/timemachine/potentials/nonbonded.py#L360

Reply:
We'd probably need to pin samples to a `DeviceBuffer` or something to truly see any meaningful performance gains here.